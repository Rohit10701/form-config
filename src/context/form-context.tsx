'use client'
import useFormWatch from '@/hooks/use-form-watch'
import React, { ReactNode, createContext, useContext, useState, useEffect } from 'react'
import { FieldValues, UseFormReturn, useWatch } from 'react-hook-form'
import { unknown } from 'zod'

export interface FormContextProps<T extends FieldValues> {
	forms: Record<string, UseFormReturn<T>>
	addForm: (id: string, methods: UseFormReturn<T>) => void
	getFormValue: (id: string, name: keyof T) => any
}

export const FormContext = createContext<FormContextProps<any>|unknown>(unknown)

export const useFormContext = <T extends FieldValues>() => {
	const context = useContext(FormContext as React.Context<FormContextProps<T> | undefined>)
	if (!context) {
		throw new Error('useFormContext must be used within a FormProvider')
	}
	return context
}

export const FormProvider = ({ children }: { children: ReactNode }) => {
	const [forms, setForms] = useState<Record<string, UseFormReturn<any>>>({})

	// Add form to the context
	const addForm = (id: string, methods: UseFormReturn<any>) => {
		setForms((prev) => ({ ...prev, [id]: methods }))
	}

	// Get form value
	const getFormValue = (id: string, name: keyof any) => {
		return forms?.[id]?.getValues(name as string)
	}


	return (
		<FormContext.Provider value={{ forms, addForm, getFormValue }}>
			{children}
		</FormContext.Provider>
	)
}
