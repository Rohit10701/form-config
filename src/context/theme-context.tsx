"use client"
import { darkTheme, lightTheme } from '@/utils/theme';
import React, { createContext, ReactNode, useContext, useState, useEffect } from 'react';

type Theme = typeof lightTheme;

interface ThemeContextProps {
  theme: Theme;
  setTheme: React.Dispatch<React.SetStateAction<Theme>>;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

export const ThemeProvider = ({ children }: { children: ReactNode }) => {
  const [theme, setTheme] = useState<Theme>(lightTheme);

  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme === lightTheme ? darkTheme : lightTheme));
  };

  return (
    <ThemeContext.Provider value={{ theme, setTheme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
