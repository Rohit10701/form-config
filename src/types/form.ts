import { FieldValuesFromFieldErrors } from '@hookform/error-message'
import React, { InputHTMLAttributes, ReactNode } from 'react'
import { FieldErrors, FieldName, Path, PathValue } from 'react-hook-form' 
import { ZodType } from 'zod'

export interface FormConfig<T extends Record<string, unknown>> {
	form: {
		id: string
		submitText: ReactNode | string
		onSubmit: (data: T) => void
	}
	// formType?: 'simple' | 'stepper' | 'otp'
	fields?: FieldInput<T>[]
	options?: {
		stepper?: StepperConfig
		otp?: OTPConfig<T>
	}

}
 export type Option = {
	label: string;
	value: string;
  };
export type FieldInput<T extends Record<string, unknown>> = InputHTMLAttributes<HTMLInputElement> & {
	name: FieldName<FieldValuesFromFieldErrors<FieldErrors<T>>>
	label?: string | ReactNode
	type: FieldType
	placeholder?: string
	style?:React.CSSProperties
	className?: string
	errors?:FieldErrors<T>
	required?: boolean
	value?:  Option | string[] | ((string | number | readonly string[]) & PathValue<T, Path<T>>) | undefined
	dependency?: {
		on: FieldName<FieldValuesFromFieldErrors<FieldErrors<T>>>[];
		condition: (value: DependencyValue<FieldName<FieldValuesFromFieldErrors<FieldErrors<T>>>[]>) => boolean;
	  };
	options?: Option[]
	validation? : unknown
	component? : React.FC<any>
	styles?: {
		container?: React.CSSProperties;
		input?: React.CSSProperties;
		label?: React.CSSProperties;
		error?: React.CSSProperties;
	  };
	customClassName?: {
		container?: string;
		input?: string;
		label?: string;
		error?: string;
	  };
	columns?: number;
	rows?: number;
}
export type DependencyValue<T extends string[]> = {
	[K in T[number]]?: string;
  };
export interface FieldConfig<T extends Record<string, unknown>> {
	fields : FieldInput<T>[]
}

// export type GenericFieldType = 'text' | 'email' | 'password' | 'number' | 'range'
// export type FieldType =
// 	| 'checkbox'
// 	| 'radio'
// 	| 'file'
// 	| 'date'
// 	| 'textarea'
// 	| 'phone'
// 	| 'select'
// 	| GenericFieldType



export type FieldType =
  | 'text'
  | 'textarea'
  | 'number'
  | 'select'
  | 'radio'
  | 'checkbox'
  | 'phone'
  | 'date'
  | 'time'
  | 'currency'
  | 'color'
  | 'file'
  | 'readonly'
  | 'url';






//not need to for prototype
interface StepperConfig {
	progress?: ReactNode
	steps: {
		title: string
		fields: string[]
		children: StepperConfig
		skip: boolean
	}[]
} 

interface OTPConfig<T extends Record<string, unknown>> {
	fields: FieldConfig<T>[]
}


export interface DynamicFormProps<T extends Record<string, unknown>> {
	id: string
	config: FormConfig<T>
	defaultValues?: Partial<T>
	schema? : ZodType<any, any, any>

}