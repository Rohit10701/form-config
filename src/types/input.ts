export type FieldType =
	| 'text'
	| 'password'
	| 'email'
	| 'number'
	| 'checkbox'
	| 'radio'
	| 'file'
	| 'date'
	| 'textarea'
	| 'label'
	| 'select'
	| 'button'
	| 'time'
	| 'phone'
	| 'autocomplete'
	| 'readonly'
	| 'tel'
	| 'url'
	| 'search'
	| 'range'
	| 'color'
	| 'datetime-local'
	| 'week'
	| 'month'
	| 'form'
	| 'fieldset'
	| 'legend'
	| 'datalist'
	| 'output'
	| 'progress'
	| 'meter'
	| 'keygen'


	