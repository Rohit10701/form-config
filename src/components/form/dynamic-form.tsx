import { useFormContext } from "@/context/form-context";
import {
  DynamicFormProps,
  FieldInput,
} from "@/types/form";
import { Controller, Path, PathValue } from "react-hook-form";
import Input from "../base/input";
import { useEffect } from "react";
import useDynamicForm from "@/hooks/use-dynamic-form";

const DynamicForm = <T extends Record<string, unknown>>({
  id,
  config,
  defaultValues,
}: DynamicFormProps<T>) => {
  const {
    control,
    handleSubmit,
    setValue,
  } = useDynamicForm<T>(id, config);

  useEffect(() => {
    if (defaultValues) {
      Object.keys(defaultValues).forEach((key) => {
        setValue(
          key as Path<T>,
          defaultValues[key as keyof T] as PathValue<T, Path<T>>,
        );
      });
    }
  }, [setValue, defaultValues]);

  const submitHandler = (data: T) => {
    config?.form?.onSubmit(data);
  };

  return (
    <>
      <form onSubmit={handleSubmit(submitHandler)}>
        {config.fields?.map((field: FieldInput<T>) => (
          <Controller
            key={field.name}
            name={field.name as Path<T>}
            control={control}
            render={({ field: controlledField }) => (
              <Input
                {...field}
                defaultValue={controlledField.defaultValue}
                onChange={controlledField.onChange}
                onBlur={controlledField.onBlur}
                disabled={controlledField.disabled}
                name={controlledField.name}
              />
            )}
          />
        ))}
        <button>{config?.form?.submitText}</button>
      </form>
    </>
  );
};

export default DynamicForm;
