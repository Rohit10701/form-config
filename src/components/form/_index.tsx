import { ReactNode, useEffect } from 'react'
import { Controller, FieldErrors, FieldName, useForm } from 'react-hook-form'
import { DependencyValue, DynamicFormProps, FieldInput } from '@/types/form'
import useDynamicForm from '@/hooks/use-dynamic-form'
import { FieldValuesFromFieldErrors } from '@hookform/error-message'
import { getFieldComponent } from '../base/_index'

const DynamicForm = <T extends Record<string, unknown>>({
	id,
	config,
	schema
}: DynamicFormProps<T>) => {
	const {
		control,
		handleSubmit,
		reset,
		watch,
		formState: { errors }
	} = useDynamicForm<T>(id, config, schema)

	// Extract default values from config fields
	useEffect(() => {
		const defaultValues = config.fields?.reduce((acc, field) => {
			// Ensure field.value exists and is included in the defaultValues
			if ('value' in field) {
				acc[field.name as keyof T] = field.value
			}
			return acc
		}, {} as T)

		if (defaultValues) {
			reset(defaultValues)
		}
	}, [config.fields, reset])

	const submitHandler = (data: T) => {
		config?.form?.onSubmit(data)
	}

	return (
		<form onSubmit={handleSubmit(submitHandler)}>
			{config?.fields?.map((fieldData: FieldInput<T>) => {
				const dependencyValues = fieldData.dependency?.on.reduce((acc, fieldName) => {
					acc[fieldName] = watch(fieldName as FieldName<FieldValuesFromFieldErrors<FieldErrors<T>>>)
					return acc
				}, {} as DependencyValue<string[]>)
				const FieldComponent = getFieldComponent(fieldData.type)

				return (
					(!fieldData.dependency ||
						(fieldData.dependency && fieldData.dependency.condition(dependencyValues))) && (
						<Controller
							key={fieldData.name}
							name={fieldData.name}
							control={control}
							render={({ field: controlledField }) => {
								const CustomComponent = fieldData?.component

								return (
									<>
										{CustomComponent ? (
											<CustomComponent
												{...fieldData}
												errors={errors}
												{...controlledField}
												onChange={controlledField.onChange as (...event: any[]) => void}
												name={
													controlledField.name as FieldName<
														FieldValuesFromFieldErrors<FieldErrors<T>>
													>
												}
											/>
										) : (
											<FieldComponent
												{...fieldData}
												errors={errors}
												{...controlledField}
												onChange={controlledField.onChange as (...event: any[]) => void}
												name={
													controlledField.name as FieldName<
														FieldValuesFromFieldErrors<FieldErrors<T>>
													>
												}
											/>
										)}
									</>
								)
							}}
						/>
					)
				)
			})}
			<button type='submit'>{config?.form?.submitText}</button>
		</form>
	)
}

export default DynamicForm
