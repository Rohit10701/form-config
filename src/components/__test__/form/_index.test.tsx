import DynamicForm from '@/components/form/_index'
import { FormProvider } from '@/context/form-context'
import { FormConfig } from '@/types/form'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { describe } from 'node:test'
import { z } from 'zod'

// Mock form configurations
type FormValueType = {
	"username" : string
	"email" : string
}
const formConfig : FormConfig<FormValueType> = {
	form: {
		id: 'test-form',
		submitText: 'Submit',
		onSubmit: jest.fn()
	},
	fields: [
		{
			name: 'username',
			label: 'Username',
			type: 'text',
			required: true,
			placeholder: 'Enter username'
		},
		{ name: 'email', label: 'Email', type: 'email', required: true, placeholder: 'Enter email' }
	]
}

// Mock schema
const schema = z.object({
	email: z.string().email(),
	email2: z.string().email().optional().or(z.literal('')),
	username: z.string().min(4).optional()
})

describe('Dynamic Form Component', () => {
	/* Rendering */
	test('should throw an error when config does not contain fields', () => {
		const renderWithoutFields = () => {
			render(
				<FormProvider>
					<DynamicForm id="1" config={{ form : {"id" : "1", "onSubmit" : ()=>{},"submitText" : ""}, fields : []}} />
				</FormProvider>
			)
		}

		expect(renderWithoutFields).toThrowErrorMatchingSnapshot('Fields are required in the config!')
	})

	test('render the component with correct config', () => {
		render(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={formConfig}
					schema={schema}
				/>
			</FormProvider>
		)

		// fields are rendered properly
		expect(screen.getByPlaceholderText('Enter username')).toBeInTheDocument()
		expect(screen.getByPlaceholderText('Enter email')).toBeInTheDocument()
		expect(screen.getByText('Submit')).toBeInTheDocument()
	})

	test('rendering of select component', () => {
		// other component is being rendered correctly apart from generic input
		const selectFormConfig : FormConfig<FormValueType & {selectField : string}> = {
			...formConfig,
			fields: [
				...formConfig.fields,
				{
					name: 'selectField',
					label: 'Select Field',
					type: 'select',
					options: [
						{ label: 'Option 1', value: 'option1' },
						{ label: 'Option 2', value: 'option2' },
						{ label: 'Option 3', value: 'option3' }
					],
					placeholder: 'slec',
					required: true
				}
			]
		}
		render(
			<FormProvider>
				<DynamicForm
					config={selectFormConfig}
					id='test-form'
				/>
			</FormProvider>
		)

		expect(screen.getByText('slec')).toBeInTheDocument()
		fireEvent.mouseDown(screen.getByText('slec'))
		fireEvent.click(screen.getByText('Option 1'))
		expect(screen.getByText('Option 1')).toBeInTheDocument()
	})

	test('toggles dark mode correctly', () => {
		const { rerender } = render(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={formConfig}
					schema={schema}
					darkMode={false}
				/>
			</FormProvider>
		)

		// expect dark mode to be off initially
		expect(screen.getByTestId('test-form')).not.toHaveClass('dark')

		// Re-render with darkMode enabled
		rerender(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={formConfig}
					schema={schema}
					darkMode={true}
				/>
			</FormProvider>
		)

		// Expect dark mode to be applied
		expect(screen.getByTestId('test-form')).toHaveClass('dark')
	})

	/* Field Dependencies &  Dynamic Behavior */
	test('test for the dependency rendering', () => {
		const dependentFormConfig : FormConfig<FormValueType & {email2 : string}> = {
			...formConfig,
			fields: [
				...formConfig.fields,
				{
					name: 'email2',
					label: 'Email2',
					type: 'email',
					required: false,
					placeholder: 'Enter email 2',
					dependency: {
						on: ['email'],
						condition: (value) => value.email === 'trigger@example.com'
					}
				}
			]
		}
		render(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={dependentFormConfig}
					schema={schema}
				/>
			</FormProvider>
		)

		// until dependency condition is true email2 input field won't get rendered
		expect(screen.queryByPlaceholderText('Enter email 2')).toBeNull()

		// enter the email "trigger@example.com" in email field
		fireEvent.change(screen.getByPlaceholderText('Enter email'), {
			target: { value: 'trigger@example.com' }
		})
		expect(screen.getByPlaceholderText('Enter email')).toHaveValue('trigger@example.com')
		expect(screen.getByPlaceholderText('Enter email 2')).toBeInTheDocument()
	})

	/* Form Submission */
	test('submit form data correctly', async () => {
		render(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={formConfig}
					schema={schema}
				/>
			</FormProvider>
		)

		// simulate user input

		fireEvent.change(screen.getByPlaceholderText('Enter username'), {
			target: { value: 'JohnDoe' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter email'), {
			target: { value: 'john@example.com' }
		})

		// user submission
		fireEvent.click(screen.getByText('Submit'))
		await waitFor(() => {
			expect(formConfig.form.onSubmit).toHaveBeenCalled()
		})
		expect(formConfig.form.onSubmit).toHaveBeenCalled()
		expect(formConfig.form.onSubmit).toHaveBeenCalledWith({
			username: 'JohnDoe',
			email: 'john@example.com'
		})
		expect(formConfig.form.onSubmit).not.toHaveBeenCalledWith({
			username: 'John Doi',
			email: 'john@example.com'
		})
	})

	test('submission with invalid data', async () => {
		render(
			<FormProvider>
				<DynamicForm
					id='test-form'
					config={formConfig}
					schema={schema}
				/>
			</FormProvider>
		)
		expect(screen.getByPlaceholderText('Enter username')).toBeInTheDocument()
		expect(screen.getByPlaceholderText('Enter email')).toBeInTheDocument()
		fireEvent.change(screen.getByPlaceholderText('Enter username'), {
			target: { value: 'JohnDoe' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter email'), { target: { value: 'a' } })

		fireEvent.click(screen.getByText('Submit'))
		await waitFor(() => {
			expect(formConfig.form.onSubmit).toHaveBeenCalled()
		})

		expect(formConfig.form.onSubmit).not.toHaveBeenCalledWith({
			username: '',
			email: ''
		})
	})

	/* Validating Schema */
	test('check for validating required schema with correct data', async () => {
		const dependentFormConfig : FormConfig<FormValueType & {email2 : string}> = {
			...formConfig,
			fields: [
				...formConfig.fields,
				{
					name: 'email2',
					label: 'Email2',
					type: 'email',
					required: false,
					placeholder: 'Enter email 2'
				}
			]
		}

		render(
			<FormProvider>
				<DynamicForm
				id="form-test"
					schema={schema}
					config={dependentFormConfig}
				/>
			</FormProvider>
		)

		fireEvent.change(screen.getByPlaceholderText('Enter email'), {
			target: { value: 'test@example.com' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter email 2'), {
			target: { value: 'r@gmail.com' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter username'), { target: { value: 'rohit' } })

		fireEvent.click(screen.getByText('Submit'))
		await waitFor(() => {
			expect(formConfig.form.onSubmit).toHaveBeenCalled()
		})

		expect(formConfig.form.onSubmit).toHaveBeenCalledWith({
			username: 'rohit',
			email: 'test@example.com',
			email2: 'r@gmail.com'
		})

		jest.clearAllMocks()

		fireEvent.change(screen.getByPlaceholderText('Enter email'), {
			target: { value: 'test@example.com' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter email 2'), { target: { value: '' } })
		fireEvent.change(screen.getByPlaceholderText('Enter username'), { target: { value: 'rohit' } })

		fireEvent.click(screen.getByText('Submit'))
		await waitFor(() => {
			expect(formConfig.form.onSubmit).toHaveBeenCalled()
		})

		expect(formConfig.form.onSubmit).toHaveBeenCalledWith({
			username: 'rohit',
			email: 'test@example.com',
			email2: ''
		})

		jest.clearAllMocks()

		fireEvent.change(screen.getByPlaceholderText('Enter email'), {
			target: { value: 'test@example.com' }
		})
		fireEvent.change(screen.getByPlaceholderText('Enter email 2'), { target: { value: '' } })
		fireEvent.change(screen.getByPlaceholderText('Enter username'), { target: { value: 'ro' } })

		fireEvent.click(screen.getByText('Submit'))

		// validtion will fail and onsubmit wont get triggred!
		await waitFor(() => {
			expect(formConfig.form.onSubmit).not.toHaveBeenCalled()
		})
	})
})
