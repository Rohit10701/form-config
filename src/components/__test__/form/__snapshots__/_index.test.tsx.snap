// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw an error when config does not contain fields: Fields are required in the config! 1`] = `"Fields are required in the config!"`;
